src/components/ActionSheetAwareScrollView/index.ios.js

```diff
@@ -0,0 +1,527 @@
+ * but maintaining the offset based on the keyboard state shared value
+ * 7. we close the picker and send EMOJI_PICKER_CLOSE action which transitions us back into keyboardOpen state.
+ *
+ * State machine object example:
+ * const stateMachine = {
+ *   idle: {
+ *       KEYBOARD_OPEN: 'keyboardOpen',
+ *   },
+ *   keyboardOpen: {
+ *       KEYBOARD_CLOSE: 'idle',
+ *       EMOJI_PICKER_OPEN: 'emojiPickerOpen',
+ *   },
+ *   emojiPickerOpen: {
+ *       EMOJI_PICKER_CLOSE: 'keyboardOpen',
+ *   },
+ * }
+ *
+ * Initial state example:
+ * {
+ *     previous: null,
+ *     current: {
+ *         state: 'idle',
+ *         payload: null,
+ *     },
+ * }
+ *
+ * @param {Object} stateMachine - a state machine object
+ * @param {Object} initialState - the initial state of the state machine
+ * @returns {Object} - an object containing the current state, a transition function, and a reset function
+ */
+function useStateMachine(stateMachine, initialState) {
+    const currentState = useSharedValue(initialState);
+
+    const transitionWorklet = useWorkletCallback((action) => {
+        if (!action) {
+            throw new Error('state machine action is required');
+        }
+
+        const state = currentState.value;
+
+        log(`Current STATE: ${state.current.state}`);
+        log(`Next ACTION: ${action.type}`, action.payload);
+
+        const nextMachine = stateMachine[state.current.state];
+
+        if (!nextMachine) {
+            log(`No next machine found for state: ${state.current.state}`);
+            return;
+        }
+
+        const nextState = nextMachine[action.type];
+
+        if (!nextState) {
+            log(`No next state found for action: ${action.type}`);
+            return;
+        }
+
+        // eslint-disable-next-line no-nested-ternary
+        const nextPayload = typeof action.payload === 'undefined' ? state.current.payload : nextState === state.current.state ? {
+            ...state.current.payload,
+            ...action.payload,
+        } : action.payload;
+
+        log(`Next STATE: ${nextState}`, nextPayload);
+
+        currentState.value = {
+            previous: state.current,
+            current: {
+                state: nextState,
+                payload: nextPayload,
+            },
+        };
+    });
+
+    const resetWorklet = useWorkletCallback(() => {
+        log('RESET STATE MACHINE');
+        currentState.value = initialState;
+    }, []);
+
+    const reset = useCallback(() => {
+        runOnUI(resetWorklet)();
+    }, []);
+
+    const transition = useCallback((action) => {
+        runOnUI(transitionWorklet)(action);
+    }, []);
+
+    return {
+        currentState,
+        transitionWorklet,
+        transition,
+        reset,
+        resetWorklet,
+    };
+}
+
+const ActionSheetAwareScrollViewContext = createContext();
+
+const stateMachine = {
+    idle: {
+        POPOVER_OPEN: 'popoverOpen',
```

Comment:
I think you should extract a separate const called `STATES`:

```js
// All the valid states of the "Action Sheet"
const STATES = {
    IDLE: 'idle',
    POPOVER_OPEN: 'popoverOpen',
    KEYBOARD_OPEN: 'keyboardOpen',
    // etc...
};
```

Then create a second const called `ACTIONS`:

```
// A set of actions that can affect the state of the "Action Sheet"
const ACTIONS = {
    OPEN_POPOVER: 'openPopover',
    OPEN_KEYBOARD: 'openKeyboard',
    CLOSE_KEYBOARD: 'closeKeyboard',
    // etc...
};
```

Then create a third object called `STATE_MACHINE`:

```js
// A state machine which represents the valid actions that can be taken from each state,
// and the ending state after those actions are taken
const STATE_MACHINE = {
    [STATES.IDLE]: {
        [ACTIONS.OPEN_POPOVER]: STATES.POPOVER_OPEN,
        [ACTIONS.OPEN_KEYBOARD]: STATES.KEYBOARD_OPEN,
        // etc...
    },
    [STATES.KEYBOARD_OPEN]: {
        [ACTIONS.CLOSE_KEYBOARD]: STATES.IDLE,
        // etc...
    },
    // etc...
};
```

Reply:
Yeah, it was my goal, but i was easier to go without it. But now it's used.